Add page on local search to documentation (#220)
Happy-Algorithms-League/hal-cgp: docs/details/local_search.rst

==========================================================
Searching numerical values for parameters via local search
==========================================================

hal-cgp supports the use of numerical values in expression via nodes with input-independent output in order to represent expressions like `f(x) = x + c` with some fixed value `c`.

> While the `ConstantFloat` node has a fixed output value, the output value of a `Parameter` node is stored per individual and can be modified during evolution via local search. These parameters can be different for each position in the genome.

Reviewer: ```suggestion
While the `ConstantFloat` node has a fixed output value, the output value of a `Parameter` node is stored per individual and can be modified during evolution via local search. These parameters can be different for each position in the genome. One can define custom nodes which contain an arbitrary number of parameters (see `example_parametrized_nodes.py`)
```
I think it would be worth making the link to the example here

Local search means that in every generation, we add an additional step where we perform a local optimization of each individual to obtain a modified invidivual. We then evaluate the fitness of the modified individual for selection in the evolutionary algorithm.

.. image:: local_search.svg

> The library provides a local search function (:meth:`cgp.local_search.gradient_based`) that performs stochastic gradient descent on the values of `Parameter` nodes (cf. Topchy & Punch, 2001; Izzo et al., 2017).

Reviewer: ```suggestion
The library provides a local search function (:meth:`cgp.local_search.gradient_based`) that performs stochastic gradient descent on the values of `Parameter` nodes (cf. Topchy & Punch, 2001; Izzo et al., 2017). Note that this method requires the definition of a differentiable loss function 
```

--------
Examples
--------

> See :ref:`sphx_glr_auto_examples_example_differential_evo_regression.py` for an example evolving an expression containing learnable parameters.

Reviewer: ```suggestion
See :ref:`sphx_glr_auto_examples_example_differential_evo_regression.py` for an example evolving an expression with numerical constants.
```

Author: I find constants here a bit misleading. "Constants" could be interpreted as if they are constant throughout the evolution.

Reviewer: well, i actually meant constants here bc i though "an expression" relates to the target function :D maybe something like "free parameters"? i just don't like "learnable" in this context, my prior is too strong ;)

Author: Yes, I know what you mean but in a way the term "constant" is overloaded in this context:
- constant across the evolution
- constant = independent from the input to the computational graph.

So I'd rather not use that word here. Why not "learnable"?

Reviewer: sorry, no, the term is _not_ overloaded with regards to the target function. that is a fixed function with some numerical values and those are generally referred to as constants. but i understand that it's more tricky if we talk about "the things in the expressions represented by individuals that are adapted by local search". so let's stick with our nomenclature and call it "parameter".

i would avoid "learnable" since for me this immediately sets the context to plasticity, brain, backprop, and so on. but these parameters can be adapted in whatever way the user wants. random search, grid search, really any continuous optimizer would be suitable and i would find it a stretch to call all of this "learning". what about "adjustable"?

Author: It is overloaded in the context of this document because just above we are referring to `ConstantNode` which is exactly what we don't mean here.

"Adjustable" is fine for me.

> See :ref:`sphx_glr_auto_examples_example_parametrized_nodes.py` for an example evolving an expression containing operator nodes with learnable parameters.

Reviewer: ```suggestion
See :ref:`sphx_glr_auto_examples_example_parametrized_nodes.py` for an example evolving an expression with numerical constants using custom defined operator node with parameters.
```

Author: same here.

> See :ref:`sphx_glr_auto_examples_example_differential_evo_regression.py` for an example evolving an expression containing learnable parameters.

Reviewer: ```suggestion
See :ref:`sphx_glr_auto_examples_example_differential_evo_regression.py` for an example evolving an expression containing adjustable parameters.
```

> See :ref:`sphx_glr_auto_examples_example_parametrized_nodes.py` for an example evolving an expression containing operator nodes with learnable parameters.

Reviewer: ```suggestion
See :ref:`sphx_glr_auto_examples_example_parametrized_nodes.py` for an example evolving an expression containing operator nodes with adjustable parameters.
```
